### DISCUSSION TITLE
Micronaut 4 JPA question: No session currently bound to execution context

### DISCUSSION THREAD

**Discussion by ethanhere** [micronaut-projects/micronaut-data#2844] [Q&A] [answered]
I have a micronaut 3 application and trying to migrate to micronaut 4 (io.micronaut.platform:micronaut-platform 4.3.5, io.micronaut:micronaut-core-bom 4.3.5)

In application.yml:

> jpa:
  default:
    packages-to-scan:
      - 'my app.entity'
    properties:
      hibernate:
        current_session_context_class: managed
        hbm2ddl:
          auto: none
        show_sql: true
        format_sql: true

The Dao class:
public class DaoImpl {

    @PersistenceContext
    protected EntityManager db;

    @Transactional
    @Override
    public void create(MyappEntity entity) {
        try {
            db.persist(entity);
            postPersist(entity);

        } catch (Exception e) {
            e.printStackTrace();
        }
    }

From Unit test, the call will get to Controller then DaoImpl.craate

Caused by: org.hibernate.HibernateException: No session currently bound to execution context
	at org.hibernate.context.internal.ManagedSessionContext.currentSession(ManagedSessionContext.java:59)
	at org.hibernate.internal.SessionFactoryImpl.getCurrentSession(SessionFactoryImpl.java:642)
	at io.micronaut.configuration.hibernate.jpa.TransactionalSessionInterceptor.intercept(TransactionalSessionInterceptor.java:62)
	at io.micronaut.aop.chain.MethodInterceptorChain.proceed(MethodInterceptorChain.java:138)
	at io.micronaut.configuration.hibernate.jpa.TransactionalSession$Intercepted.persist(Unknown Source)
        at Myapp.DaoImpl.create(DaoImpl.java:63)

I tried many combination but cannot get it working. Can someone share some insights how to fix it?

**Comment by graemerocher** [comment]
perhaps remove `current_session_context_class: managed`

**Reply by ethanhere** [reply]
@graemerocher If I remove "current_session_context_class: managed", the error will become 
caused by: org.hibernate.internal.SessionFactoryImpl.getCurrentSession(SessionFactoryImpl.java:640) 
org.hibernate.HibernateException: No CurrentSessionContext configured
	at org.hibernate.internal.SessionFactoryImpl.getCurrentSession(SessionFactoryImpl.java:640)
	at io.micronaut.configuration.hibernate.jpa.TransactionalSessionInterceptor.intercept(TransactionalSessionInterceptor.java:62)
	at io.micronaut.aop.chain.MethodInterceptorChain.proceed(MethodInterceptorChain.java:138)
	at io.micronaut.configuration.hibernate.jpa.TransactionalSession$Intercepted.persist(Unknown Source)
	at myapp.DaoImpl.create(DaoImpl.java:63)

**Reply by ethanhere** [reply]
In org.hibernate.internal.SessionFactoryImpl.java:

	String CURRENT_SESSION_CONTEXT_CLASS = "hibernate.current_session_context_class";

	private CurrentSessionContext buildCurrentSessionContext() {
		final String sessionContextType = (String) settings.get( CURRENT_SESSION_CONTEXT_CLASS );
		// for backward-compatibility
		if ( sessionContextType == null ) {
			if ( canAccessTransactionManager() ) {
				return createSessionContext( "jta" );
			}
			else {
				return null;
			}
		}
		else {
			return createSessionContext( sessionContextType );
		}
	}

So if I specify "current_session_context_class: managed" in application.yaml, then It will  createSessionContext( "managed" ), otherwise it will createSessionContext( "jta" ). Either way it's going to fail:
without current_session_context_class, the error is "No CurrentSessionContext configured"
if "current_session_context_class: managed", the error is "No session currently bound to execution context"
if "current_session_context_class: thread", the error is "org.hibernate.HibernateException: Calling method 'persist' is not valid without an active transaction (Current status: NOT_ACTIVE)"
If "current_session_context_class: jta", the error is:
10:22:59.404 [virtual-executor1] INFO  o.h.e.t.j.p.i.JtaPlatformInitiator - HHH000489: No JTA platform available (set 'hibernate.transaction.jta.platform' to enable JTA platform integration)
10:22:59.405 [virtual-executor1] ERROR m.DaoImpl - Error creating audit event
org.hibernate.HibernateException: No TransactionManagerLookup specified
	at org.hibernate.context.internal.JTASessionContext.currentSession(JTASessionContext.java:69)
	at org.hibernate.internal.SessionFactoryImpl.getCurrentSession(SessionFactoryImpl.java:642)

The above current_session_context_class values were inspired from SessionFactoryImpl.java code:
	private CurrentSessionContext createSessionContext(String sessionContextType) {
		switch ( sessionContextType ) {
			case "jta":
//			if ( ! transactionFactory().compatibleWithJtaSynchronization() ) {
//				LOG.autoFlushWillNotWork();
//			}
				return new JTASessionContext(this);
			case "thread":
				return new ThreadLocalSessionContext(this);
			case "managed":
				return new ManagedSessionContext(this);
			default:
				try {
					return (CurrentSessionContext)
							serviceRegistry.getService( ClassLoaderService.class )
									.classForName( sessionContextType )
									.getConstructor( new Class[]{ SessionFactoryImplementor.class } )
									.newInstance( this );
				}
				catch ( Throwable t ) {
					LOG.unableToConstructCurrentSessionContext( sessionContextType, t );
					return null;
				}
		}
	}

**Comment by graemerocher** [comment]
it should be https://github.com/micronaut-projects/micronaut-data/blob/942d1cb423593b297e27c3a222880313c5e7eed2/data-tx-hibernate/src/main/java/io/micronaut/transaction/hibernate/MicronautSessionContext.java

**Reply by ethanhere** [reply]
thanks for the hint. I use maven, what pom dependency contains MicronautSessionContext ?  any other dependencies? micronaut source in GitHub uses Gradle, I like to know how to figure out the dependency artifact from the java file URL.
After I include the dependency that contains MicronautSessionContext in the pom, what else do I need to do in order make the SessionFactoryImpl to create MicronautSessionContext as current session context?

**Reply by graemerocher** [reply, marked as answer]
try set `current_session_context_class: io.micronaut.transaction.hibernate.MicronautSessionContext`

**Reply by ethanhere** [reply]
Thanks, that works. 
Just set it in application.yml under
jpa.default.properties.hibernate
